---
output: pdf_document
---

# Population genetics workshop


### Welcome 

This excercise is going to expose you to several basic ideas in probability and statistics as well as show you the utility of using R for basic statistical analyses.  We'll do so in the context of a basic population genetic analysis.  

### The scenario

Your first question whenever you get access to new data should be: "Is this data clean?  Do the observations that have been made seem reliable or are they full of noise and error?"  To convince yourself you can trust a data-set it's useful to carry out some basic exploratory analyses with the data to show they conform to well-known expectations.  As a biologist, you will learn what are the major patterns that are expected when the data you work with is clean. Using that expertise will save you from the mistake of misinterpreting error-prone data.  

In population genetics, there are a number of patterns that we expect to see immediately in our datasets.  In this exercise you will explore one of those major patterns.  Rather than give it a way - let's begin some analysis and see what we find.  In the narative that follows, we'll refine our thinking.  


### Introductory terminology

- Single-nucleotide polymorphism (SNP):  A nucleotide basepair that is *polymorphic* (i.e. it has multiple types or *alleles* in the population)
- Allele:  A particular variant form of DNA  (e.g. A partciular SNP may have the "A-T" allele in one DNA copy and "C-G" in another; We typically define a reference strand of the DNA to read off of, and then denote the alleles according to the reference strand base - so for example, these might be called simply the "A" and "C" alleles.  In many cases we don't care about the precise base, so we might call these simply the $A_1$ and $A_2$ alleles, or the $A$ or $a$ alleles, or the 0 and 1 alleles.) 
- Minor allele: The allele that is more rare in a population 
- Major allele: The allele that is more common in a population 
- Genotype: The set of alleles carried by an individual (E.g. AA, AC, CC; or AA, AA, and aa; or 0/0, 1/1, 2/2)
- Genotyping array: A technology based on hybridization with probes and florensence that allows genotype calls to be made at 100s of thousands of SNPs per individual at an affordable cost.  

### The data-set and basic pre-processing

We will look at Illumina 650Y genotyping array data from the CEPH-Human Genome Diversity
Panel.  This sample is a global-scale sampling of human diversity with
52 populations in total.

The data were first described in Li et al (Science, 2008) and the raw files are available from the
following link:
<http://hagsc.org/hgdp/files.html>.  The genotypes were filtered with a GenCall score cutoff of 0.25 (a quality score generated by the basic genotype calling software).  Individuals with a genotype call rate <98.5% were removed, with the logic being that if a sample has many missing genotypes it may due to poor quality of the source DNA, and so none of the genotypes should be trusted.  Beyond this, to prepare the data for the workshop, we have filtered down the indiviudals to a set of 938 unrelated individuals.  (For those who are interested, the data are available as plink-formatted files `H938.bed` 
`H938.fam`, `H938.bim` from this link: <http://bit.ly/1aluTln>). 

We have done additional pre-processing of the data such that you have files that contain genotype frequencies. 

### Note about logistics

If you need the files they can be found at:
[[https://github.com/StefanoAllesina/BSD-QBio2/tree/master/workshops/novembre]]

### Initial view of the data

We will use some of the dplyr functions so first let's load the library:
```{r}
library(dplyr)
library(ggplot2)
```
If you get an error message you may need to first run ```install.packages("dplyr")```

Read in the data table:
```{r}
g<-read.table("../data/H938_chr2.geno",header=T)
```
It will be read in as a dataframe in R.
And use the "head" command to see the beginning of the dataframe:
```{r}
head(g)
```

You should see that there are columns each with distinct names.  
```
CHR        SNP A1 A2 nA1A1 nA1A2 nA2A2
```
- CHR: The chromosome number.  In this case they are all from chromosome 2.
- SNP: The rsid of a SNP is a unique identifier for a SNP and you can use the rsid to look up information about a SNP using online resource such as dbSNP or SNPedia. 
- A1: The minor allele at the SNP
- A2: The major allele
- nA1A1 : The number of A1/A1 homozygotes
- nA1A2 : The number of A1/A2 homozygotes
- nA2A2 : The number of A2/A2 homozygotes

### Calculate the number of counts at each locus

Compute the total number of observations, and then print a summary of the distribution:
```{r}
g <- mutate(g, nObs=nA1A1+nA1A2+nA2A2)
summary(g$nObs)
```
Run `head(g)` and confirm your dataframe `g` has a new column called `nObs`.

And let's visualize the distribution using `qplot` to produce a quick histogram:
```{r}
qplot(nObs,data=g)
```

Our data are from 938 individuals. When the counts are less than this total, it's because some individuals had array data that was difficult to call a genotype for.  

*Question:*  Do most of the SNPs have complete data?  
*Question:* What is the lowest count observed?   Is this number in rough agreement with what we know about how the genome-wide missingness rate filter was set to >98.5% of all SNPs 

(Answers: Yes; 884 is lower count.  884/938 is missingness rate fo 94.2%, lower than genome-wide average but not distressingly so)

### Calculating genotype and allele frequencies

Let's move on to calculating genotype and allele frequencies.
```{r}
# Compute genotype frequencies
g <- mutate(g, p11 = nA1A1/nObs , p12 = nA1A2/nObs, p22 = nA2A2/nObs )
# Compute allele frequencies from genotype frequencies
g <- mutate(g, p1 = p11 + 0.5*p12, p2 = p22 + 0.5*p12)
```

Run `head(g)` again and confirm `g` now has the extra columns for the genotype and allele frequencies. 

And let's plot the frequency of the major allele (A2) vs the frequency of the minor allele (A1):
```{r}
qplot(p1,p2,data=g)
```

As expected, 'p2>p1', and there is a linear relationship.  

*Question:*  What is the equation decribing this relationship?

The relationship exists because there are only two alleles - and so their proportions must sum to 1.  That is p1+p2=1, which implies the linear relationship you've found (if they didn't sum to 1 it would suggest somethings is wrong with our code!).  

### Plotting genotype on allele frequencies

Let's look at an initial plot of genotype vs allele frequencies:
```{r}
ggplot(g)+geom_point(aes(x=p1,y=p11),color="red")+
  geom_point(aes(x=p1,y=p12),color="orange")+
  geom_point(aes(x=p1,y=p22),color="blue")+
  scale_color_manual(values=c("p11"="red","p12"="orange","p22"="blue"))
```
** TODO fix legends and axis labels... I'm better with base plotting functions

There are some very striking relationships here.    

Given there are so many data points, we can plot an estimated probability density to make out the data more clearly
```{r}
ggplot(g)+geom_density2d(aes(x=p1,y=p11),color="red")+
  geom_density2d(aes(x=p1,y=p12),color="orange")+
  geom_density2d(aes(x=p1,y=p22),color="blue")
```

*Question:* What are approximate relationships between p11 vs p1, p12 vs p1, and p22 vs p1?  (Hint: These look like parabolas, which suggests are some very simple quadratic functions of p1). 

Notice these are the classic relationships you likely learned in your intro biology courses. These are the Hardy-Weinberg proportions of $p^2$,$2pq$, and $q^2$.  On average, the data follow these classic theoretical expectations fairly well.  By eye, we can see some deficiency of heterozygotes and excess of homozygotes, but in general the equations are roughly right!

Let's look more closely and more formally though... 

### Testing Hardy Weinberg

We can test Hardy-Weinberg at each SNP using Pearson's $\chi^2$-test, and using the $\chi^2$-test statistic:
$$ X^2=\sum_i \frac{(o_i-e_i)^2}{e_i}$$.  Here we compute the test statistic and obtain its associated p-value, keeping in mind that there is 1 degree of freedom (because we have 3 observations per SNP, and we lose an extra degree of freedom because the expectations depend on the data through the estimated allele frequency).

```{r}
g<-mutate(g,chisq = (p11-p1^2)^2/p1^2+(p12-2*p1*p2)^2/(2*p1*p2)+(p22-p2^2)^2/p2^2)
g<-mutate(g,pval = pchisq(chisq,1))
```

### The problem of multiple testing

Normally we would take a p-value of less than  0.05 as a sign that a SNP departs from Hardy-Weinberg expectations.  This is problematic though.  A p-value gives us the frequency at which that the observed departure from expectations (or a more extreme departure) would occur under if the null hypothesis is true.  As an agred upon standard, if the data are relatively rare under the null (i.e. p-value < 5%), we reject the null hypothesis.  The problem is that under the null (in our case HW proportions are true), we would expect a p-value < 0.05, 5% of the time.  In the classical terms, this is our expected rate of Type I error.   A consequence of all this is that is that, under the null, when we have 53,652 SNPs,  we should expect to see ~2683 tests that would have p-values < 0.05.  We clearly need more stringent approaches, that is we need to come up with corrections to deal with the "multiple testing problem".  Two frameworks are the Bonferroni approach and false-discovery-rate (FDR) based approaches.  We will not say more about these here.  In two simple ways, let's see though if our data are globally consistent with the null.  

First, let's see how many tests have p-values less than 0.05.  Is it much larger than the 2683 we'd expect on average given 53,652 SNPs and a 5% type I error rate?  

```{r}
sum(g$pval<0.05,na.rm=TRUE)
```

Wow - we see many more.  This is our first sign that though by eye these data are close to HW, there is some underlying signal that they are not precisely fitting Hardy-Weinberg. 

Let's look at this another way.  A classic result from Fisher is that under the null hypothesis the p-values of a well-designed test should be distributed uniformly between 0 and 1.  What do we see here? 

```{r}
qplot(pval,data=g)
```

The data show an enrichment for small p-values relative to a uniform distribution.  Notice how the whole distribution is shifted towards small values - The data appear to systematically depart from Hardy-Weinberg.  

### Plotting expected vs observed heterozygosity

To understand this more clearly, let's make a quick plot of the expected vs observed heterozygosity:
```{r}
qplot(2*p1*(1-p1),p12,data=g)
```
Most of the points fall below the y=x line.  That is we see a systematic deficiency of heterozygotes (and this implies a concordant excess of homozygotes).  This general pattern is contributing to the departure from HW seen in the $X^2$ statistics. 

### Discussion: Population subdivision and departures from Hardy-Weinberg expectations

We might wonder why the deaprture from Hardy-Weinberg proportional is directional, in that, on average, we are seeing a deficiency of heterozygotes.  One enlightening way to understand this is by thinking about what Sewall Wright called "the correlation of uniting gametes".  To produce an $A_1 A_1$ individual we need an $A_1$-bearing sperm and an $A_1$-bearing egg to unite.  If these events were independent of each other, we would expect $A_1 A_1$ individuals at the rate predicted by multiplying probabilites, that is, $p_1^2$.  However, what if uniting gametes are positively correlated, in that an A-bearing sperm is more likely to joint with an A-bearing egg?  In this case we will have more $A_1 A_1$ individuals than predicted by $p_1^2$, and conversely fewer $A_1 A_2$ individuals than predicted by $p_1 p_2$.  If our population is structured somehow such that $A_1$ sperm are more likely to meet with $A_1$ eggs, then we will have such a positive correlation of uniting gametes, and the resulting excess of homozygotes and deficiency of heterozygotes. 

Given the HGDP data is from 52 sub-populations from around the globe, and alleles have some some probability of clustering within poulations, a good working hypotheiss for the deficiency of heterozygotes in this dataset is the presence of some population structure.  

While statistically significant, the population structure appears to be subtle --- based on our first plots, we saw the genotype proportions are not wildly off from HW proportions. How can a global sample be relatively close in absolute terms to a random mating population?  Partly because human populations are all relatively recently derived from a single source population in Africa. Most of the alleles in the sample are widespread and not geographically clustered in ways that generate correlations among the uniting alleles.   

*Question:* As an exercise, compute the average deficiency of heterozygotes relative to the expectated proportion.  That is the average of $(2p_1 p_2-p_{12})/2p_1 p_2)$.  What is the number?  A common "rule-of-thumb" for this value in a global sample of humans is that the deficiency is approximately 10\%.  Do you find this to be true from the data? 

### Finding specific loci that are large departures from Hardy-Weinberg

Now, let's ask if we can find any loci that are wild departures from HW proportions.  These might be loci that have erroneous genotypes, or loci that cluster geographically in dramatic ways (such that they have few heterozygotes relative to expectations).

To find these loci, we'll compute the same relative deficiency you computed above.  That is:
$$ \frac{2 p_1 (1-p_1)-p_{12}}{2p_1(1-p_1)}$$
This number (1) is referred to as the $F$ by Sewall Wright, (2) has connections directly to correlation coefficients (advanced exercise: Try to work this out!), and (3) if we assume there is no inbreeding within populations, this is related to $F_{ST}$ (which you may have seen).

Let's add this value to our dataframe and plot a histogram:

```{r}
g<-mutate(g,F=(2*p1*(1-p1)-p12)/(2*p1*(1-p1)))
qplot(F,data=g)
```

Now, here's a trick.  When a high $F$ is due to genotyping error, it likely only effects a single SNP.  However, when there is some population genetic force acting on a region of the genome, it likely effects multiple SNPs in the region.  So let's use a function to take a running mean of SNPs across the genome, computing an average over every 5 consecutive SNPs (in real data analysis we might use 100kb or 0.1cM windows). 

```{r}
movingavg <- function(x,n=5){stats::filter(x,rep(1/n,n), sides=2)}
plot(movingavg(g$F))
```

Wow - there appears to be one large spike in the dataset. 

Let's extract the SNP id for the largest value:
```{r}
g$SNP[which (movingavg(g$F) == max(movingavg(g$F)))]
```

*Question:* Which SNP id is returned?  By inserting the rsid into the UCSC genome browser (https://genome.ucsc.edu/), and following the links, find out what gene this SNP resides near.  What gene is it?

*Question:* Carry out a literature search on this gene using the term "positive selection" and see what you find. It's thought the high $F$ value observed here is because natural selection led to a geographic clustering of alleles in this gene region.  A recent publication in Cell is particularly interesting in this regard. 


### Discussion: The outlier reigon

The region you've found is one of the most differentiated betwen human populations that is known.  Notice in your literature search, how it is known to affect morphological phenotypes such as hair thickness and is thought to explain differences in average hair thickness that are seen between human populations.  This is not uncommon when looking at other chromosomes.  Some of the most differentiated genes that exist in humans are those that involve morphological phenotypes - such as skin pigmentation, hair color/thickness, and eye color (the genes SLC45A2, KITLG, OCA2/HERC2, SCL24A5 all come to mind).  Many of these are thought to have arisen due to direct or indirect effects of adaptation to local selective pressures (e.g. adaptation to varing levels of UV exposure, local pathogens, local diets, local mating preferences), though in most cases we still do not yet have a fully convincing understanding of their evolutionary histories.  Regardless of the reasons, it is notable that many of the features that humans see externally in each other (i.e. the morphological differences) are controlled by genes that are outliers in the genome.  At most variant SNPs, the patterns of variation are much closer to those of a single random mating populations than they are at variant sites like EDAR.  Put another way, a genomic perspective shows us many of the differences people see in each other are in a sense, just skin-deep. 

### Wrap-up

Modern population genetics has a lot of additional tools on its workbench, but here using relatively simple and classical ideas combined with genomic-scale data, we have been able to observe and interpret some major features of human genetic diversity.  We have seen how at a global scale human populations show about a 10% departure on average from the expectations of Hardy-Weinberg and that at particular regions, like EDAR, there are stronger departures.  We have also revisited some basic concepts of probablity and statistics such as indepedence vs correlation, the $\chi^2$ test, and the problems of multiple testing.  Finally, we have gained more familiarity with R. In the end, we didn't do much about the possible genotyping errors (those regions where there are isolated SNPs that depart from Hardy-Weinberg), but if we were doing additional analyses, we could filter them from out of the data. 


### Follow-up activities

In the 'addons' folder, we are including data files that you can use to explore.  These include global data for other chromosomes (`H938_chr*.geno`) and the same data but limited to European populations (`H938_Euro_chr*.geno`). 

*Follow-up actvity*: Look at the data from European populations - is the global deficiency in heterozygosity as strong as it was on the global scale?  What would you expect to see?  

*Follow-up activity*: Using the global data or the European data, analyze other chromosomes -- do you find other loci that show high $F$ values?

*Follow-up activity* Using the European data, do you find any regions of the genome that are outliers for $F$ on chromosome 2?  What genes are in that region?
